orttraining/orttraining/core/graph/zero_optimizer_graph_builder.cc

```diff
@@ -93,39 +95,97 @@ static Status AddL2NormNcclAllReduce(
   return Status::OK();
 }
 
+static std::string GetViewName(const std::string param_name, const size_t view_num) {
+  return param_name + "_view_" + std::to_string(view_num);
+}
+
+static std::vector<ArgDef> AddPartitionsForParameter(
+    Graph& graph,
+    GraphAugmenter::GraphDefs& graph_defs,
+    const std::string& initializer_name,
+    const std::vector<TensorShape>& shapes,
+    std::unordered_map<std::string, std::string>& updated_weight_names_map){
+  ORT_ENFORCE(shapes.size() == 3, "Invalid shapes vector passed for partitioning.");
+  int64_t partition_offset = shapes[0].GetDims()[0];
+  int64_t partition_size = shapes[1].GetDims()[0];
+  std::vector<ArgDef> view_outputs;
+  for (size_t i = 0, view_num = 0; i < shapes.size(); i++) {
+    if (shapes[i].Size() > 0) {
+      // keep the naming consistent with the "Views" created for the rest of the parameters.
+      const auto partition_name = GetViewName(initializer_name, view_num);
+      if (i == 1) {
+        const ONNX_NAMESPACE::TensorProto* tensor_proto;
+        ORT_ENFORCE(graph.GetInitializedTensor(initializer_name, tensor_proto));
+        const auto dtype = static_cast<ONNX_NAMESPACE::TensorProto_DataType>(tensor_proto->data_type());
+        // Shard only fp32 weight for zero stage 1
+        ORT_ENFORCE(dtype == ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+
+        // Find the initializer partition to read out.
+        auto initializer = onnxruntime::make_unique<Initializer>(*tensor_proto, graph.ModelPath());
+        const float* initializer_data = initializer->data<float>();
+
+        // Create new initializer tensor proto.
+        ONNX_NAMESPACE::TensorProto initializer_partition;
+        initializer_partition.set_name(partition_name);
+        initializer_partition.set_data_type(dtype);
+        initializer_partition.add_dims(partition_size);
+        initializer_partition.set_raw_data(initializer_data + partition_offset, partition_size * sizeof(float));
+
+        //Replace the old initializer with the new one
+        graph.RemoveInitializedTensor(initializer_name);
+        graph.AddInitializedTensor(initializer_partition);
+
+        //add the modified weight name to get state 
+        updated_weight_names_map.insert({initializer_name, partition_name});
```

Comment:
one comment here since I don't know the intent behind this statement. Just wanted to make sure that the intent here is not to overwrite the element in the map but rather insert only if the element does not exist in the map, right?

Reply:
the first scenario doesnt exist here, but changing in any case